### PR TITLE
Output error when forward referencing constants in inline assembly

### LINKED ISSUE #8881: ICE in const solidity::frontend::TypePointer &solidity::frontend::TypeChecker::type(const solidity::frontend::Expression &) const
## Description

This contract:
```solidity
contract C {
    function o (int256) public returns (int256) {
	assembly {
	    c:=shl(1,a)
	}
    }

    int constant c=2 szabo+1 seconds+3 finney*3 hours;
}
```

produces:

```
Internal compiler error during compilation:
/home/user/solidity/libsolidity/analysis/TypeChecker.cpp(74): Throw in function const solidity::frontend::TypePointer &solidity::frontend::TypeChecker::type(const solidity::frontend::Expression &) const
Dynamic exception type: boost::exception_detail::clone_impl<solidity::langutil::InternalCompilerError>
std::exception::what: Type requested but not present.
[solidity::util::tag_comment*] = Type requested but not present.
```

when compiled with solc

On master, using AFL fuzzing.  Another discovery using https://github.com/agroce/afl-compiler-fuzzer

Looks similar to some closed bugs...

## Environment

- Compiler version:  0.6.8-develop.2020.5.9+commit.4e58c672.Linux.clang
- Target EVM version (as per compiler settings): N/A
- Framework/IDE (e.g. Truffle or Remix): N/A
- EVM execution environment / backend / blockchain client: N/A
- Operating system: Ubuntu 18.04 in docker

## Steps to Reproduce

Above shows pretty clearly, I think.

### PR DESCRIPTION
**Pull request by Marenz**
fix #8881

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -709,6 +701,23 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 					m_errorReporter.typeError(6617_error, _identifier.location, "The suffixes _offset and _slot can only be used on non-constant storage variables.");
 					return size_t(-1);
 				}
+				else if (var && var->value() && !var->value()->annotation().type)
+				{
+					m_errorReporter.typeError(
+						2249_error,
+						_identifier.location,
+						"Constant variables cannot be forward referenced."
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
						"Constant variables cannot be forward referenced from inline assembly."
```

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: Actually if it is a literal, it should be fine.

Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp]: Isn't it a literal in this case?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: No, it is an expression

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -709,6 +701,23 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 					m_errorReporter.typeError(6617_error, _identifier.location, "The suffixes _offset and _slot can only be used on non-constant storage variables.");
 					return size_t(-1);
 				}
+				else if (var && var->value() && !var->value()->annotation().type)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Please make this work for direct literal values.

File: test/libsolidity/syntaxTests/inlineAssembly/const_forward_reference.sol [hunk 3]
```diff
@@ -0,0 +1,8 @@
+contract C {
+  function f() public pure {
+    assembly {
+      pop(add(shl(1, 2), c))
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/inlineAssembly/const_forward_reference.sol]: shl is not always available.

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -709,6 +701,23 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 					m_errorReporter.typeError(6617_error, _identifier.location, "The suffixes _offset and _slot can only be used on non-constant storage variables.");
 					return size_t(-1);
 				}
+				else if (var && var->value() && !var->value()->annotation().type && dynamic_cast<Literal const*>(var->value().get()) == nullptr)
```
Comment (hrkrshnn) [inline libsolidity/analysis/TypeChecker.cpp]: What do you think about:
```suggestion
				else if (var && var->value() && !var->value()->annotation().type && !dynamic_cast<Literal const*>(var->value().get()))
```

Comment (hrkrshnn) [reply libsolidity/analysis/TypeChecker.cpp]: The file already has `== nullptr` pattern. So maybe this change is not needed. Core guidelines recommend removal: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es87-dont-add-redundant--or--to-conditions

Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp]: Changed

**Review by chriseth — APPROVED** [review]

**Pull request merged**